src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -3498,21 +3498,55 @@ void StorageReplicatedMergeTree::removePartAndEnqueueFetch(const String & part_n
         removePartFromZooKeeper(part_name, ops, stat.numChildren > 0);
     }
 
-    LogEntryPtr log_entry = std::make_shared<LogEntry>();
-    log_entry->type = LogEntry::GET_PART;
-    log_entry->create_time = part_create_time;
-    log_entry->source_replica = "";
-    log_entry->new_part_name = part_name;
 
-    ops.emplace_back(zkutil::makeCreateRequest(
-        fs::path(replica_path) / "queue/queue-", log_entry->toString(),
-        zkutil::CreateMode::PersistentSequential));
+    while (true)
+    {
+        /// We use merge predicate + version check here, because DROP RANGE update log version and we are trying to avoid race with it. We must be sure, that our part
+        /// was not dropped, otherwise we will have fetch entry, but no virtual part for it (DROP RANGE will remove it). So bad sequence is the following:
+        /// 1) Create DROP PART in log for broken_part (for example because it's not only broken, but also empty)
+        /// 1.5) Pull to log and start execution (removePartProducingOpsInRange)
+        /// 2) Execute this method (create GET PART for broken part)
+        /// 3) Finish DROP PART execution (remove part from virtual parts)
+        /// 4) Now we get GET PART for broken part which is absent in virtual parts == bug
+        ///
+        /// Because of version check this method will never create FETCH if drop part exists
+        ReplicatedMergeTreeMergePredicate merge_pred = queue.getMergePredicate(zookeeper, PartitionIdsHint{broken_part_info.partition_id});
+        if (merge_pred.hasDropRange(broken_part_info))
+        {
+            LOG_INFO(log, "Broken part {} is covered by DROP RANGE, don't need to fetch it", part_name);
+            return;
+        }
+
+        LogEntryPtr log_entry = std::make_shared<LogEntry>();
+        log_entry->type = LogEntry::GET_PART;
+        log_entry->create_time = part_create_time;
+        log_entry->source_replica = "";
+        log_entry->new_part_name = part_name;
 
-    auto results = zookeeper->multi(ops);
+        /// Check that our version of log (and queue) is the most fresh. Otherwise don't create new entry fetch entry.
+        ops.emplace_back(zkutil::makeCheckRequest(fs::path(zookeeper_path) / "log", merge_pred.getVersion()));
```

Comment:
the check from the previous loop will stay in the current and fail multi request again?

Reply:
Absolutely. Stupid bug, thanks.